File: Tests/CorePromise/ErrorUnhandler.test.m
#import <PromiseKit/PromiseKit.h>
@import XCTest;


@interface WTFError : NSError @end @implementation WTFError
@end

@interface NSError (BridgingHack)
+ (void)pmk_setUnhandledErrorHandler:(void (^)(NSError *))block;
@end

void PMKSetUnhandledErrorHandler(void (^block)(NSError *)) {
    [NSError pmk_setUnhandledErrorHandler:block];
}


@interface ErrorHandlingTests_ObjC: XCTestCase @end @implementation ErrorHandlingTests_ObjC

- (void)tearDown {
    PMKSetUnhandledErrorHandler(^(id err){});
}

- (void)test_68_unhandled_error_handler {
    @autoreleasepool {
        XCTestExpectation *ex = [self expectationWithDescription:@""];

        PMKSetUnhandledErrorHandler(^(NSError *error){
            XCTAssertEqual(error.code, 5);
            XCTAssertEqualObjects(@"a", error.domain);
            [ex fulfill];
        });

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            resolve([NSError errorWithDomain:@"a" code:5 userInfo:@{@1: @2}]);
        }];
    }
    [self waitForExpectationsWithTimeout:2 handler:nil];
}


- (void)test_69_unhandled_handled_returned {
    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@"unhandler"];
        XCTestExpectation *ex2 = [self expectationWithDescription:@"initial catch"];
        PMKSetUnhandledErrorHandler(^(NSError *error){
            XCTAssertEqual(5, error.code);
            [ex1 fulfill];
        });
        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            resolve([NSError errorWithDomain:@"a" code:5 userInfo:nil]);
        }].catch(^(id e){
            [ex2 fulfill];
            return e;
        }).then(^{
            XCTFail();
        });
    }
    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_70_unhandled_error_handler_not_called {
    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@""];

        PMKSetUnhandledErrorHandler(^(id err){
            XCTFail();
        });

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            resolve([NSError errorWithDomain:@"a" code:5 userInfo:nil]);
        }].catch(^{
            return dispatch_promise(^{
                return dispatch_promise(^{
                    @throw @"5";
                });
            });
        }).catch(^{
            [ex1 fulfill];
        });
    }

    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_79_unhandled_error_handler_not_called_reject_passed_through {
    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@""];

        PMKSetUnhandledErrorHandler(^(id e){
            XCTFail();
        });

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            dispatch_promise(^{
                @throw @"1";
            }).catch(resolve);
        }].catch(^{
            [ex1 fulfill];
        });
    }

    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_80_unhandled_error_handler_called_if_reject_passed_through {
    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@""];
        XCTestExpectation *ex2 = [self expectationWithDescription:@""];

        __block BOOL ex1Fulfilled = NO;

        PMKSetUnhandledErrorHandler(^(id e){
            XCTAssert(ex1Fulfilled);
            [ex2 fulfill];
        });

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            dispatch_promise(^{
                @throw @"1";
            }).catch(resolve);
        }].finally(^{
            [ex1 fulfill];
            ex1Fulfilled = YES;
        });
    }
    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_999_allow_error_subclasses {
    XCTestExpectation *ex1 = [self expectationWithDescription:@""];

    PMKAfter(0.02).then(^{
        return [WTFError errorWithDomain:@"WTF" code:0 userInfo:nil];
    }).catch(^(NSError *e){
        XCTAssertEqualObjects(e.class, WTFError.class);
        [ex1 fulfill];
    });

    [self waitForExpectationsWithTimeout:1 handler:nil];
}

@end

File: Tests/CorePromise/ErrorUnhandler.test.swift
import Foundation
import XCTest
import PromiseKit

private enum Error: ErrorProtocol {
    case dummy
}

class ErrorHandlingTests_Swift: XCTestCase {

    var oldHandler: ((ErrorProtocol) -> Void)!

    override func setUp() {
        oldHandler = PMKUnhandledErrorHandler
    }
    override func tearDown() {
        PMKUnhandledErrorHandler = oldHandler
    }

    private func twice(body: @noescape (Promise<Int>, XCTestExpectation) -> Void) {
        autoreleasepool {
            let ex = expectation(withDescription: "Sealed")
            body(Promise<Int>.resolved(error: Error.dummy), ex)
        }
        waitForExpectations(withTimeout: 1, handler: nil)

        autoreleasepool {
            let ex = expectation(withDescription: "Unsealed")
            let p = Promise { fulfill, _ in
                fulfill(1)
            }.then { (_: Int) -> Int in
                throw Error.dummy
            }
            body(p, ex)
        }
        waitForExpectations(withTimeout: 1, handler: nil)
    }

    // handler is called if error is not handled
    func test1() {
        twice { promise, ex in
            PMKUnhandledErrorHandler = { err in
                ex.fulfill()
            }
        }
    }

    // handler is *not* called if error is caught
    func test2() {
        twice { promise, ex in
            PMKUnhandledErrorHandler = { err in
                XCTFail()
            }
            promise.catch { error in
                ex.fulfill()
            }
        }
    }

    // handler is *not* called if error is recovered
    func test3() {
        twice { promise, ex in
            PMKUnhandledErrorHandler = { err in
                XCTFail()
            }
            _ = promise.recover { error -> Promise<Int> in
                return Promise.resolved(value: 1)
            }.always {
                ex.fulfill()
            }
        }
    }

    // handler is *still* called if error is *not* recovered
    func test4() {
        twice { promise, ex in
            PMKUnhandledErrorHandler = { err in
                ex.fulfill()
            }
            _ = promise.recover { error -> Int in
                throw error
            }
        }
    }

    // handler is called *once* if recovery fails with a *different* error
    func test5() {
        twice { promise, ex in
            PMKUnhandledErrorHandler = { err in
                ex.fulfill()
            }
            promise.recover { error -> Promise<Int> in
                firstly { throw Error.dummy }
            }
        }
    }

    // handler is *not* called if error recovery fails with same error but eventually error is caught
    func test6() {
        twice { promise, ex2 in
            let ex1 = expectation(withDescription: "")

            PMKUnhandledErrorHandler = { err in
                XCTFail()
            }
            promise.recover { error -> Int in
                throw error
            }.then { x in
                XCTFail()
            }.always {
                ex1.fulfill()
            }.catch { err in
                ex2.fulfill()
            }
        }
    }

    func testDoubleRejectDoesNotTriggerUnhandler() {
        enum Error: ErrorProtocol {
            case test
        }

        PMKUnhandledErrorHandler = { err in
            XCTFail()
        }

        let (p, _, r) = Promise<Void>.pending()

        let ex1 = expectation(withDescription: "")
        let ex2 = expectation(withDescription: "")
        let ex3 = expectation(withDescription: "")
        let ex4 = expectation(withDescription: "")

        after(interval: 0.1).then { _ -> Void in r(Error.test); ex1.fulfill() }
        after(interval: 0.15).then { _ -> Void in r(Error.test); ex2.fulfill() }.always(execute: ex3.fulfill)

        p.catch { error in
            ex4.fulfill()
        }

        waitForExpectations(withTimeout: 1, handler: nil)
    }

    func testPassThrough() {
        let ex = expectation(withDescription: "")

        PMKUnhandledErrorHandler = { err in
            ex.fulfill()
        }

        enum Error: ErrorProtocol {
            case test
        }

        Promise<Void> { _, reject in
            after(interval: 0.1).then {
                throw Error.test
            }.catch(execute: reject)
        }

        waitForExpectations(withTimeout: 1, handler: nil)
    }

    func testConsumedPromiseStaysConsumedAsAnyPromise() {
        enum Error: ErrorProtocol {
            case test
        }

        PMKUnhandledErrorHandler = { err in
            XCTFail()
        }

        let ex1 = expectation(withDescription: "")

        let p: Promise<Int> = firstly {
            throw Error.test
        }

        XCTAssertTrue(p.isRejected)

        let anyp = AnyPromise(bound: p)

        p.catch { err in
            ex1.fulfill()
        }

        waitForExpectations(withTimeout: 1, handler: nil)

        print(anyp)
    }
}


extension NSError {
    @objc class public func pmk_setUnhandledErrorHandler(_ handler: (NSError) -> Void) {
        PMKUnhandledErrorHandler = { (error: ErrorProtocol) -> Void in
            handler(error as NSError)
        }
    }
}
